### DISCUSSION TITLE
Duplicate custom clusters and herdsman

### DISCUSSION THREAD

**Discussion by sjorge** [Koenkk/zigbee2mqtt#9574] [General]
cc @Koenkk

So I was looking through more of the compliance documents for the IKEA STARKVIND and found this:
`0x042a	 PM2.5 Concentration Measurement client(yes) server(no)`

Which does seem to match: https://github.com/Koenkk/zigbee-herdsman/blob/e0b6ceaeca24ede57930f2552b2332b7108ca211/src/zcl/definition/cluster.ts#L4127-L4139

After poking around some more, I found 0x042a is indeed a custom pm25 cluster for the EmberZNet SDK, which seems to be what IKEA is using as the same document mentions: `EmberZNet 6.7.5.0`.

When trying to use the existing definition, I get UNSUPPORTED_ATTRIBUTE. I tried adding a copy but with the IKEA manufacturing code:

```node
    ikeaSpecificPM25Measurement: {
        ID: 0x042a,
        manufacturerCode: ManufacturerCode.IKEA_OF_SWEDEN,
        attributes: {
            measuredValue: {ID: 0x0000, type: DataType.uint16},
            measuredMinValue: {ID: 0x0001, type: DataType.uint16},
            measuredMaxValue: {ID: 0x0002, type: DataType.uint16},
            measuredTolerance: {ID: 0x0003, type: DataType.uint16},
        },
        commands: {},
        commandsResponse: {},
    },
```

I pushed and installed the git based zigbee-herdsman on my install in both the main zigbee2mqtt and zigbee-herdsman-converters dependency.

Looks like the frontend is at least aware of them and I can use them in the configure function, but they don't seem to work and herdsman is now very confused.

e.g. if I switch to normal dev (no changes) I can bind `heimanSpecificPM25Measurement` in the frontend.
If I switch back with my zigbee-herdsman with the extra definition I cannot unbind or bind `heimanSpecificPM25Measurement`.

Switching back to unmodified I was able to unbind `heimanSpecificPM25Measurement` again just fine.

With my modified herdsman install I am also unable to read heimanSpecificPM25Measurement or ikeaSpecificPM25Measurement, I think it might just be getting confused somewhere.

At the very least, I would expect to get a UNSUPPORT_ATTRIBUTE back when trying to use ikeaSpecificPM25Measurement.

So what's the proper way to add the same cluster but with 2 manufacture variants? Perhaps there should just be one generic PM25Measurement definition in herdsman that will automatically fill in the correct manufacturereCode based on the device it is discovered on? With the option to override it via options (I think that already works?)

**Comment by Koenkk** [comment]
Have you checked that the manufacturercode is required in the first place?

**Reply by sjorge** [reply]
> Have you checked that the manufacturercode is required in the first place?

I did try that, but as far as I can see it's getting confused and trying to use the heimanSpecificPM25Measurement regardless.
I also found a newer PICS document and it seems those are client clusters only so it's not like we can get the info from them :(

Still no access to a gateway sadly.

**Reply by Koenkk** [reply]
I don't understand why zigbee-herdsman gets confused, when configure reporting it should use the name which is unique. Can you provide the debug log (when using modified herdsman with both clusters) when configuring reporting?

**Reply by sjorge** [reply]
I figured it out -_- it's a client cluster on the device, so of course you cant bind/configure reporting on it.

**Comment by sjorge** [comment]
@Koenkk related question... given I figured out the particular cluster was client side, but it did show up on my dev z2m install I just did so I probably messed it up first. I also added the other 2 unknown clusters that do show up on the as server! One even responds, now to figure out WHAT it responds with.

```
Zigbee2MQTT:debug 2021-11-10 21:09:12: Received MQTT message on 'zigbee2mqtt/airpurifier/bedroom/set' with data '{"read":{"attributes": ["measuredValue"],"cluster":"ikeaSpecificFC7D","options":{}}}'
Zigbee2MQTT:debug 2021-11-10 21:09:13: Received Zigbee message from 'airpurifier/bedroom', type 'readResponse', cluster 'ikeaSpecificFC7D', data '{"measuredValue":2273}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-11-10 21:09:13: No converter available for 'E2007' with cluster 'ikeaSpecificFC7D' and type 'readResponse' and data '{"measuredValue":2273}'
Zigbee2MQTT:info  2021-11-10 21:09:13: Read result of 'ikeaSpecificFC7D': {"measuredValue":2273}
```

I noticed neither of those 2 server clustered showed up in the frontend before I added them.
<img width="984" alt="image" src="https://user-images.githubusercontent.com/379665/141194216-1df62e5b-324c-41b1-be8c-445d50d0daf5.png">

Would it be possible to show those if there is no configuration in herdsman as 'Unknown 0xXXXX' ? That way it's easy to see which devices have currently unknown clustered and if they are client or server based on the position in the tree. I'm not sure this is a herdman feature request or a frontend one.

**Reply by sjorge** [reply]
On that note, I am def on to something with this one.

```
Zigbee2MQTT:info  2021-11-10 21:23:22: Read result of 'manuSpecificIkeaPM25Measurement': {"measuredValue":2287,"measuredMinValue":0,"measuredMaxValue":259200,"measuredTolerance":0}
```

Looks like it's very very likely the PM25 sensor! will this poke more.

**Reply by Koenkk** [reply]
> Would it be possible to show those if there is no configuration in herdsman as 'Unknown 0xXXXX' ?

Done, in this case it will show the number (see blue one below):

<img width="757" alt="Screenshot 2021-11-11 at 21 04 43" src="https://user-images.githubusercontent.com/2892853/141361756-8285fd68-b191-4a90-a0bb-b9d7c5e1f959.png">

**Reply by sjorge** [reply]
Nice, decimal numbers is fine 64637 or 0xfc7d doesn't really make a difference. It's just nice to see them now atleast.

**Reply by hakaneriksson** [reply]
Just got my ZBOSS sniffer up. 
When starting the device it say on app "Air quality: Collecting info...". Then comes the following frames and it says "Good".
```
ZigBee Cluster Library Frame, Mfr: Ikea of Sweden (0x117c), Command: Report Attributes, Seq: 84
    Frame Control Field: Profile-wide (0x0c)
    Manufacturer Code: Ikea of Sweden (0x117c)
    Sequence Number: 84
    Command: Report Attributes (0x0a)
    Attribute Field, Uint16: 3
        Attribute: 0x0004
        Data Type: 16-Bit Unsigned Integer (0x21)
        Uint16: 3 (0x0003)

ZigBee Cluster Library Frame, Mfr: Ikea of Sweden (0x117c), Command: Report Attributes, Seq: 85
    Frame Control Field: Profile-wide (0x0c)
    Manufacturer Code: Ikea of Sweden (0x117c)
    Sequence Number: 85
    Command: Report Attributes (0x0a)
    Attribute Field, Uint16: 8
        Attribute: 0x0004
        Data Type: 16-Bit Unsigned Integer (0x21)
        Uint16: 8 (0x0008)
```
